components/sup/src/manager/service/spawned_future.rs

```diff
@@ -0,0 +1,56 @@
+// Copyright (c) 2019 Chef Software Inc. and/or applicable contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Encapsulate a pattern where we spawn a task on a separate thread,
+//! but need to represent the failure to spawn that thread, as well as
+//! the return value of that thread's computation as a single Future.
+//!
+//! It's essentially a tailored version `futures::FutureResult`,
+//! except that it polls its `Ok` variant.
+
+use crate::error::SupError;
+use futures::{sync::oneshot, Future, Poll};
+use std::io;
+
+pub enum SpawnedFuture<T> {
+    Ok(oneshot::Receiver<T>),
+    Err(Option<io::Error>),
+}
```

Comment:
I think you may be right that this can be simplified a bit. I was following the example of [FutureResult](https://docs.rs/futures/0.1.25/futures/future/struct.FutureResult.html) which uses an `Option` (and `take`) as a way of ensuring a particular future can only be called once. (I've seen this pattern elsewhere, as well.)

This was early-ish code, and once I got it working, I didn't come back to refactor. I'll take another look.

Reply:
Upon further reflection, I'd like to leave this as it is for now. The main thing that `SpawnedFuture` is doing is allowing us to handle the case where there's some kind of error in spawning the thread will generate the `oneshot::Receiver`. If the spawning fails, then we generate the error variant, which will get polled and immediately return that error. If the spawning is successful, then we'll end up with the `Ok` variant, which just holds onto the real future that we poll as usual.

Creating a type to handle both cases made it easier to encapsulate that and express what `HookRunner` and the service terminator were doing.

Reply:
Oh geez... yeah, oversight. Fixing!